server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs

```diff
@@ -0,0 +1,440 @@
+
+    -- We can also ignore the websocket 'CloseRequest' exception because wai-websockets ensures
+    -- that the connection is closed when this IO action finishes, as this IO action is executing in a bracket.
+    -- https://github.com/yesodweb/wai/blob/c92d7b1993338fb0f91e0f5f34fb9678871028a0/wai-websockets/Network/Wai/Handler/WebSockets.hs#L98
+    handleConnectionError :: WS.ConnectionException -> IO ()
+    handleConnectionError _ = pure ()
+
+    -- Since we are in a bracket that closes the connection, let the action silently exit.
+    handleSubscriptionTimeout :: SubscriptionTimeout -> IO ()
+    handleSubscriptionTimeout _ = pure ()
+
+  -- Put the client in the subscription tree and keep the connection open.
+  -- Remove it when the connection is closed.
+  Exception.finally
+    (onConnect client >> manageConnection)
+    (onDisconnect client)
+    `Exception.catch` handleConnectionError
+    `Exception.catch` handleSubscriptionTimeout
+
+-- After timeout, throw the action an exception if the client hasn't subscribed.
+withSubscribeTimeout :: Client -> IO a -> IO a
+withSubscribeTimeout client action = do
+  let
+    initalSubscriptionTimeout =
+      Config.configInitialSubscriptionTimeoutMicroSeconds $
+        Core.coreConfig $
+          clientCore client
+
+    clientNoSubscribers = do
+      let subscriptions = clientSubscriptions client
+      HashMap.null <$> MVar.readMVar subscriptions
+
+  threadId <- Concurrent.myThreadId
+  Monad.void $ Concurrent.forkIO $ do
+    Concurrent.threadDelay initalSubscriptionTimeout
+    noSubscribers <- clientNoSubscribers
+    Monad.when noSubscribers (Concurrent.throwTo threadId SubscriptionTimeout)
+  action
+
+startMessageHandlerThread :: Client -> IO ()
+startMessageHandlerThread client = Monad.forever $ do
+  let
+    coreConfig = Core.coreConfig $ clientCore client
+    conn = clientConn client
+  -- Note: WS.receiveDataMessage will handle control messages automatically.
+  -- Upon a close control message, a CloseRequest exception will be thrown.
+  dataMessage <- WS.receiveDataMessage conn
+  onPayload coreConfig client $
+    parseDataMessage dataMessage
+
+takeMVarUpdatedValues :: Client -> IO [(Text, Value)]
+takeMVarUpdatedValues client = do
+  let
+    isDirty = clientIsDirty client
+    subscriptions = clientSubscriptions client
+  MVar.takeMVar isDirty
+  valueMVars <- HashMap.toList <$> MVar.readMVar subscriptions
+  Maybe.catMaybes
+    <$> ( Monad.forM valueMVars $
+            \(path, valueMVar) ->
+              fmap (\v -> (Text.intercalate "/" path, v))
+                <$> MVar.tryTakeMVar valueMVar
+        )
+
+-- This function handles sending the updates to subscribers.
+startUpdaterThread :: Client -> IO ()
+startUpdaterThread client =
+  let
+    conn = clientConn client
+
+    send :: Text -> Value -> IO ()
+    send path value =
+      Exception.catch
+        ( WS.sendTextData conn $
+            Aeson.encode $
+              Update path value
```

Comment:
Perhaps already answered elsewhere, but is the websocket sending thread safe? (I.e. when the updater thread decides to send an update message at the same time as the payload handler sends a response to a request, does the client receive two separate valid messages, or do they end up garbled? In the former, we obviously don't have control on the order, but we also don't need to, I'm just concerned about the integrity of the messages individually.)

Reply:
I did not consider this 😳

Reply:
Looks like it is thread safe, it uses a lock, says the comment:
- https://hackage.haskell.org/package/websockets-0.13.0.0/docs/src/Network.WebSockets.Stream.html#makeStream